### PR TITLE
Combine query builder & index apis

### PR DESCRIPTION
**Pull request by ricellis**
- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://github.com/cloudant/java-cloudant/blob/master/DCO1.1.txt)
- [x] You have added tests for any code changes
- [x] You have updated the [CHANGES.md](https://github.com/cloudant/java-cloudant/blob/master/CHANGES.md) - these changes already recorded from previous merges
- [x] You have completed the PR template below:

## What

Common up parts of the `QueryBuilder` and `Index` improvements.

## How

* Added statics to construct `Sort`
* Refactored `QueryBuilder` to use `Sort`
* Renamed `OperationOrExpression` -> `Selector`
* Used a single `Type` enum instead of having two versions

## Testing

Refactored the new tests to accommodate these changes.

## Issues

N/A

### CODE REVIEW

File: cloudant-client/src/main/java/com/cloudant/client/internal/util/SelectorUtils.java [hunk 1]
```diff
@@ -17,18 +17,89 @@
...
+import com.cloudant.client.api.query.Selector;
 import com.google.gson.Gson;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
 
 public class SelectorUtils {
 
-    public static JsonObject getPartialFilterSelectorFromString(String selectorJson) {
-        return getSelectorFromString("partial_filter_selector", selectorJson);
+    // Selector object keys
```
Comment (tomblench) [inline cloudant-client/src/main/java/com/cloudant/client/internal/util/SelectorUtils.java]: I wonder whether this has overlap with my `Helpers` class

File: cloudant-client/src/main/java/com/cloudant/client/api/query/Selector.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*
```
Comment (tomblench) [inline cloudant-client/src/main/java/com/cloudant/client/api/query/Selector.java]: Did you remove `OperationOrExpression.java`? I couldn't see it in the diff but not sure if it's just a github quirk.

Comment (ricellis) [reply cloudant-client/src/main/java/com/cloudant/client/api/query/Selector.java]: Good spot, I thought I'd done a rename, but it looks like I somehow kept the original. Fixed in 3c4aa85cdc3ec79f01851f02027f3af454c2d09b

File: cloudant-client/src/main/java/com/cloudant/client/internal/util/SelectorUtils.java (hunk 1 above)
Comment (ricellis) [reply cloudant-client/src/main/java/com/cloudant/client/internal/util/SelectorUtils.java]: Moved `SelectorUtils` methods into `Helpers`, unfortunately it didn't look like I could re-use any of the methods. See b2c9a25a46f7754a1a681737a1694b5b0dd0c6eb

**Comment by ricellis** [comment]
Also moved the Predicate classes to the internal package in b4a7072ebbb9e57c9d802367925b76fb6b06a108

**Pull request merged**